Sources/ServiceLifecycle/ServiceGroup.swift

```diff
@@ -126,32 +147,46 @@ public actor ServiceGroup: Sendable {
     }
 
     private enum ChildTaskResult {
-        case serviceFinished(service: any Service, index: Int)
-        case serviceThrew(service: any Service, index: Int, error: any Error)
+        case serviceFinished(service: ServiceGroupConfiguration.ServiceConfiguration, index: Int)
+        case serviceThrew(service: ServiceGroupConfiguration.ServiceConfiguration, index: Int, error: any Error)
         case signalCaught(UnixSignal)
         case signalSequenceFinished
         case gracefulShutdownCaught
         case gracefulShutdownFinished
     }
 
-    private func _run(gracefulShutdownStream: AsyncStream<Void>) async throws {
+    private func _run(
+        services: inout [ServiceGroupConfiguration.ServiceConfiguration],
+        gracefulShutdownStream: AsyncStream<Void>
+    ) async throws {
         self.logger.debug(
             "Starting service lifecycle",
             metadata: [
-                self.configuration.logging.keys.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
-                self.configuration.logging.keys.servicesKey: "\(self.services)",
+                self.loggingConfiguration.keys.gracefulShutdownSignalsKey: "\(self.gracefulShutdownSignals)",
+                self.loggingConfiguration.keys.cancellationSignalsKey: "\(self.cancellationSignals)",
+                self.loggingConfiguration.keys.servicesKey: "\(services.map { $0.service })",
             ]
         )
 
         // Using a result here since we want a task group that has non-throwing child tasks
         // but the body itself is throwing
         let result = await withTaskGroup(of: ChildTaskResult.self, returning: Result<Void, Error>.self) { group in
             // First we have to register our signals.
-            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+            let gracefulShutdownSignals = await UnixSignalsSequence(trapping: self.gracefulShutdownSignals)
+            let cancellationSignals = await UnixSignalsSequence(trapping: self.cancellationSignals)
```

Comment:
Might be simpler to just create one sequence because both tasks for these are exactly the same...

Reply:
I tried that first but it makes things a bit more complicated because right now I return from the child task on the first signal; however, we have to keep listening to cancellation signals once we get a graceful shutdown signal.